### PR TITLE
Convert names_and_types graph APIs to pybind11

### PR DESCRIPTION
**Pull request by hidmic**
Precisely what the title says. Part of #665. This patch also brings `rcpputils` as a dependency.

CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13960/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8793/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11660/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14043/)

### CODE REVIEW

File: rclpy/src/rclpy/_rclpy_pybind11.cpp [hunk 1]
```diff
@@ -226,4 +227,33 @@ PYBIND11_MODULE(_rclpy_pybind11, m) {
...
+  m.def(
+    "rclpy_get_subscriber_names_and_types_by_node",
+    &rclpy::graph_get_subscriber_names_and_types_by_node,
+    "Get topic names and types for which a given remote node has subscribers.");
+  m.def(
+    "rclpy_get_service_names_and_types",
+    &rclpy::graph_get_service_names_and_types,
+    "Get all service names and types in the ROS graph.");
+  m.def(
+    "rclpy_get_service_names_and_types",
```
Comment (sloretz) [inline rclpy/src/rclpy/_rclpy_pybind11.cpp:248]: It looks like the `rclpy_get_service_names_and_types` `m.def(` call is duplicated.

File: rclpy/src/rclpy/_rclpy_pybind11.cpp [hunk 2]
```diff
@@ -226,4 +227,33 @@ PYBIND11_MODULE(_rclpy_pybind11, m) {
...
+    &rclpy::graph_get_service_names_and_types,
+    "Get all service names and types in the ROS graph.");
```
Comment (gbalke) [inline rclpy/src/rclpy/_rclpy_pybind11.cpp:250]: Duplicate?

File: rclpy/src/rclpy/graph.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+#include "rclpy_common/common.hpp"
+#include "rclpy_common/exceptions.hpp"
+#include "rclpy_common/handle.h"
+
+#include "graph.hpp"
+
+
+namespace rclpy
+{
+
```
Comment (sloretz) [inline rclpy/src/rclpy/graph.cpp:35]: Huh, I could have sworn `cpplint` complained when there was a blank line after a namespace opening, but the test passes locally for me :shrug:

No need to change it. If the linter is happy I'm happy :upside_down_face:

File: rclpy/src/rclpy/graph.hpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLPY__GRAPH_HPP_
+#define RCLPY__GRAPH_HPP_
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
```
Comment (sloretz) [inline rclpy/src/rclpy/graph.hpp:18]: Nit, there's no handle.h include here

File: rclpy/src/rclpy/graph.hpp [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+namespace py = pybind11;
+
+namespace rclpy
+{
+
+/// Get topic names and types for which a given remote node has publishers.
+/**
+ * Raises ValueError if pynode is not a node capsule
+ * Raises RCLError if there is an rcl error
```
Comment (sloretz) [inline rclpy/src/rclpy/graph.hpp:34]: Please also add `NodeNameNonExistentError`

File: rclpy/src/rclpy_common/include/rclpy_common/common.hpp [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+namespace py = pybind11;
+
+namespace rclpy
+{
+
+/// Convert a C rcl_names_and_types_t into a Python list.
+/**
+ * Raises ValueError if pywait_set is not a wait set capsule
+ * Raises RuntimeError if the entity type is not known
+ *
```
Comment (gbalke) [inline rclpy/src/rclpy_common/include/rclpy_common/common.hpp:33]: Where is `pywait_set` coming from?

**Review by gbalke — APPROVED** [review]
Looks good! Couple comments/questions.

**Review by sloretz — CHANGES_REQUESTED** [review]
It looks like the removal of code from `_rclpy.c` is missing from the PR, mind adding it?

File: rclpy/src/rclpy/graph.hpp (hunk 5 above)
Comment (gbalke) [reply rclpy/src/rclpy/graph.hpp:34]: I was going to ask this. `NodeNameNonExistant` extends `RCLError`. Should it be separately mentioned? If so, then all of the docstrings need to be updated.

Comment (hidmic) [reply rclpy/src/rclpy/graph.hpp:34]: Done in https://github.com/ros2/rclpy/pull/717/commits/a13f583864c6a3ff3d68fbd577acb3e86361eb07.

File: rclpy/src/rclpy/graph.hpp (hunk 4 above)
Comment (hidmic) [reply rclpy/src/rclpy/graph.hpp]: Removed  in https://github.com/ros2/rclpy/pull/717/commits/a13f583864c6a3ff3d68fbd577acb3e86361eb07.

File: rclpy/src/rclpy/graph.cpp (hunk 3 above)
Comment (hidmic) [reply rclpy/src/rclpy/graph.cpp:35]: Aha! I had not noticed. `ament_uncrustify` seems to be OK with it too.

File: rclpy/src/rclpy/_rclpy_pybind11.cpp (hunk 1 above)
Comment (hidmic) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp:244]: Yeap, my bad. I wonder why `pybind11` didn't complain though. See https://github.com/ros2/rclpy/pull/717/commits/a13f583864c6a3ff3d68fbd577acb3e86361eb07.

File: rclpy/src/rclpy_common/include/rclpy_common/common.hpp (hunk 6 above)
Comment (hidmic) [reply rclpy/src/rclpy_common/include/rclpy_common/common.hpp]: Blatant copy & paste. Fixed in https://github.com/ros2/rclpy/pull/717/commits/a13f583864c6a3ff3d68fbd577acb3e86361eb07.

File: rclpy/src/rclpy/_rclpy_pybind11.cpp (hunk 2 above)
Comment (hidmic) [reply rclpy/src/rclpy/_rclpy_pybind11.cpp]: Indeed. Removed duplicate in https://github.com/ros2/rclpy/pull/717/commits/a13f583864c6a3ff3d68fbd577acb3e86361eb07.

**Comment by hidmic** [comment]
Running CI up to `rclpy`:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13964/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8796/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11664/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14047/)

**Comment by hidmic** [comment]
Rebased to solve conflicts.

**Comment by hidmic** [comment]
CI up to `rclpy`, third's the charm:

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13965/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8797/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11665/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14048/)

File: rclpy/src/rclpy_common/include/rclpy_common/common.hpp [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+/// Convert a C rcl_names_and_types_t into a Python list.
+/**
+ * \param[in] names_and_types The names and types struct to convert.
+ * \return List of tuples, where the first element of each tuple is a string
+ *   for the name and the second element is a list of strings for the types.
+ */
+RCLPY_COMMON_PUBLIC
+py::list
+convert_to_py_names_and_types(rcl_names_and_types_t * topic_names_and_types);
```
Comment (sloretz) [inline rclpy/src/rclpy_common/include/rclpy_common/common.hpp:37]: Nit, recommend making this parameter `const`.

**Review by sloretz — APPROVED** [review]
One nitpick that I wouldn't block on.

File: rclpy/src/rclpy_common/include/rclpy_common/common.hpp (hunk 7 above)
Comment (hidmic) [reply rclpy/src/rclpy_common/include/rclpy_common/common.hpp]: Done in https://github.com/ros2/rclpy/pull/717/commits/b016dbc2546442049f299d9e3640b4d76ae33b60.

**Review by sloretz — APPROVED** [review]

**Comment by sloretz** [comment]
CI with const change (build: `--packages-up-to rclpy` test: `--packages-select rclpy`)
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/13969/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/8800/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/11668/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/14052/)

**Comment by sloretz** [comment]
@hidmic, It looks like this one is ready to go! I'll merge it so it doesn't need to be rebased again if something else would have gotten merged first.

**Comment by clalancette** [comment]
I'm not 100% sure if it was this PR, but Windows debug build is now broken in rclpy: https://ci.ros2.org/view/nightly/job/nightly_win_deb/1931/ .  The only two PRs that were merged yesterday were this one and #712 .  @sloretz @hidmic can you take a look?

**Comment by hidmic** [comment]
It's trying to link to non-debug Python libraries:

```
LINK : fatal error LNK1104: cannot open file 'python38.lib' [C:\ci\ws\build\rclpy\rclpy_common.vcxproj]
```

but the CMakeLists.txt does seem to be taking care of not doing that. Has this happened to you before @sloretz ?

**Comment by sloretz** [comment]
> but the CMakeLists.txt does seem to be taking care of not doing that. Has this happened to you before @sloretz ?

Yeah, this seems familiar. There's some odd linker pragmas in the python.h and pybind11.h headers involving the `Py_DEBUG` macro. I'm guessing if we include <pybind11/pybind11.h> before `<python.h>` in the other compilation units in `rclpy_common`, then the issue would be resolved.

**Comment by sloretz** [comment]
I think we've been getting lucky with the order. I'm guessing the linker pragma set by the last compilation unit is the one that wins :-/

The jobs that passed, the C++ file comes last:

```
  Building Custom Rule C:/ci/ws/src/ros2/rclpy/rclpy/CMakeLists.txt

  common.c

  handle.c

  exceptions.cpp

     Creating library C:/ci/ws/build/rclpy/Debug/rclpy_common.lib and object C:/ci/ws/build/rclpy/Debug/rclpy_common.exp
```

The job that failed, a C file came last

```
 Building Custom Rule C:/ci/ws/src/ros2/rclpy/rclpy/CMakeLists.txt

  common.c

  common.cpp

  exceptions.cpp

  handle.c

LINK : fatal error LNK1104: cannot open file 'python38.lib' [C:\ci\ws\build\rclpy\rclpy_common.vcxproj]
```

**Comment by hidmic** [comment]
Oh my. Will fix now.

**Pull request merged**